index.rst

```diff
@@ -13,17 +13,256 @@
+
+Datastore methods that add new datasets to the repository would be modified to return a collection of records describing those datasets, again for Butler to pass to Registry.
+Datastore methods that read datasets or interpret the records describing them would be modified to accept those records from Butler (which fetches them from Registry).
+Some Datastore existence-check methods would go away entirely (e.g. ``knows``), as their functionality is subsumed by Registry dataset queries, while others would change their behavior to checking for artifact existence *given* records.
+``Registry.insertDatasets`` would be modified to accept datastore records for storage, and ``Registry.findDataset`` would be modified to return dataset records as well as a ``DatasetRef``.
+
+.. note::
+   All ``DatasetRef`` objects in this technote are assumed to be resolved; unresolved ``DatasetRef`` objects are already slated to go away per :jira:`RFC-888`.
+
+This proposal formalizes what we are already doing during no-database batch execution, while taking advantage of new developments - UUIDs and ``daf_relation`` - to simplify the Registry/Datastore boundary.
+It would involve considerable code changes, but more removals than additions, and the vast majority of these would be behind the scenes or of minimal impact to users (e.g. ``Butler.datastore`` and ``Registry.insertDatasets`` are not formally private, but they should be, and are already widely recognized as for internal use only).
+
+Implementation of important butler operations
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In order for this proposal to be considered viable, we need to look carefully at how we would implement joint Registry/Datastore operations that we want to maintain consistency.
+Note that later sections in this technical note will expand upon these and ultimately take the form of prototype Python code, as further changes are added to the proposal.
+
+``Butler.put``
+""""""""""""""
+
+1. Obtain a valid expanded data ID from the Registry.
+   In the vast majority of cases (i.e. QuantumGraph execution) this done well in advance of the actual ``put`` call.
+
+2. Construct a ``DatasetRef`` by generating an appropriate UUID and using an existing or soon-to-exist RUN name.
+   This will also typically occur well before the rest of the ``put``, as part of QuantumGraph generation.
+
+3. Perform the ``Datastore.put`` operation, writing the file artifacts associated with the dataset and returning records to the Butler.
+   Datastore can be expected to make this operation atomic, either because it is naturally atomic for its storage backing or via writing a temporary and moving it.
+   We do have to accept the possibility of failures leaving partially-written temporary files around.
+   When the Datastore is backed by storage that requires signed URLs, this write is to a staging area and uses non-signed URIs.
```

Comment:
Is the staging area client-side (but with Datastore tables updated?!) or server-side?  A client-side staging area is not terrible, as `resources` writes to local storage anyway before upload, but I'd very much like to avoid a server-side one, particularly if it uses unsigned URIs (i.e. a dumping ground for any user — or anyone on the Internet?).

Reply:
That "passing in" part reminded me that we need to make sure that we never trust anything passed in through the c/s Butler API — we must check for access before issuing URLs.  (Well, we can usually trust opaque, long, randomish numbers/strings that we gave out previously.)

Reply:
Always. This is covered a little in DMTN-242. It is assumed that all server APIs take an auth token and then check collection accessibility.

Reply:
I'm totally on board with rewriting this to have the Datastore asking for signed URLs in advance (or something equivalent; maybe Butler does it for Datastore?) instead of having a staging area.

I don't have a useful opinion on where we try to handle conflicts when signed URLs are in play.  But my previous mantra of, "it needs to be solid in terms of consistency and races, but not necessarily fast" for that conflict-handling still applies.